content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,511 @@
+
+1. Wrote an algorithm to prove that a type set is empty, thus proving that it
+   can be done efficiently. Or
+2. Wrote a proof that the problem is still NP-complete.
+
+The issue is, that doing either of these takes time. I did invest that time in
+the original discussion, before the limitations where introduced and so far, I
+did not get around to doing it with them in place. Maybe someone else has, or
+will.
+
+Personally, I'm not sure there is a huge incentive, though. If you write a
+constraint with an accidentally empty type set, even the most rudimentary
+testing will expose that, as a function using it can not be instantiated and
+thus not used. And even *if* you made such a mistake, fixing it would then not
+even be a breaking change - as no one could ever use that function anyways.
+
+I also feel that all of this hints at a deeper problem with the concepts of
+type sets. There are a couple of more or less arbitrary restrictions in place.
+Arguably, too many: The
+[design doc contains an example use case](https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#interface-types-in-union-elements)
+which [no longer works](https://go.dev/play/p/mwdmdgKg6pj).
+
+To me, such restrictions indicate a wrong approach from the start. For example,
+perhaps we could have instead made the extension to the interface syntax in a
+way that type sets are given in DNF, instead of CNF. Or perhaps we should not
+have introduced `|` in the first place.
+
+If you do have an algorithm, or a proof, feel free to let me know. I'd be happy
+to provide feedback on either.
+
+[^1]: It should be pointed out, though, that “polynomial” can still be
+  extremely inefficient. n<sup>1000</sup> still grows extremely fast, but is
+  polynomial. And for many practical problems, even n<sup>3</sup> is
+  intolerably slow. But for complicated reasons, there is a qualitatively
+  important difference between “polynomial” and “exponential”[^6] run time, so
+  just trust me, that the distinction makes sense.
+
+[^2]: These names might seem strange, by the way. P is easy to explain: It
+  stands for “polynomial”. NP doesn't mean “not polynomial”, though. It means
+  “non-deterministic polynomial”. A non-deterministic computer, in this
+  context, is a hypothetical machine which can run arbitrarily many
+  computations simultaneously. A program which can be *decided* efficiently by
+  any computer, can be *solved* efficiently by a non-deterministic one, by just
+  trying out all possible solutions at the same time and returning a correct
+  one. Thus, being able to decide a problem on a normal computer, means being
+  able to solve it on a non-deterministic one.
+
+[^3]: You might complain that it is hard to remember if the “disjunctive normal
+  form” is a disjunction of conjunctions, or a conjunction of disjunctions -
+  and that no one can remember which of these means “and” and which means “or”
+  anyways. You would be correct.
```

Comment:
😆

Reply:
I honestly have to look this up every. Single. Time.
So I'm joking, but I'm also not. This is absolute garbage jargon.
Unfortunately, I'm generally stuck with using it, I think.

Reply:
I haven't had to deal with these terms very often, but it didn't seem that bad to me?  "X normal form" seems clear that it means "a list of things joined by X": "Conjunctive Normal Form" => "a list of things joined by conjunctions". 
 `Join([]interface{}, &&)`, so to speak 😆